### PR TITLE
[HOTFIX] detach() callbacks should called even if not anchored()

### PR DESCRIPTION
**Pull request by jtlan**
`detach()` callbacks should be called whenever `detach()` is called, to make sure that the `Component` and its parent correctly disassociate.

#### Testing Notes
The bug manifested when `detach()`-ing `Component`s that were in a `Group` or `Table` before the `Group` or `Table` was rendered.

### CODE REVIEW

**Comment by aicioara** [comment]
Can we make a followup ticket where we address the problem properly? It feels weird to me to call `detach` callbacks when no actual detaching is performed.

**Comment by aicioara** [comment]
Can we add a test which covers this case

```typescript
c.onDetach(callback);

c.detach();
assert.isTrue(callbackWasCalled);

callbackWasCalled = false

c.detach();
assert.isTrue(callbackWasCalled);

```

**Comment by jtlan** [comment]
What is the suggested test case supposed to check for?

**Comment by aicioara** [comment]
I am personally unsure what is going to happen in that scenario. I don't see anything covering the scenario

**Comment by crmorford** [comment]
tables: destroy and detach http://jsfiddle.net/rzspuLvx/
groups: destroy and detach http://jsfiddle.net/rzspuLvx/1/
detach then add to group http://jsfiddle.net/rzspuLvx/2/
detach then add to table http://jsfiddle.net/rzspuLvx/3/
destroy then add to group http://jsfiddle.net/rzspuLvx/4/ (FAILS -- expected)
destroy then add to table http://jsfiddle.net/rzspuLvx/5/ (FAILS -- expected)

**Comment by crmorford** [comment]
detach after render http://jsfiddle.net/rzspuLvx/6/
detach and reattach, after render http://jsfiddle.net/rzspuLvx/7/
destroy after render http://jsfiddle.net/rzspuLvx/8/
destroy and reattach, after render http://jsfiddle.net/rzspuLvx/9/ (FAILS -- expected)
